lib/Mojolicious/Guides/Testing.pod

```diff
@@ -0,0 +1,754 @@
+
+  use Test::Mojo;
+  use Test::More;
+
+  # Start a Mojolicious app named "Celestial"
+  my $t = Test::Mojo->new('Celestial');
+
+  # Post a JSON document
+  $t->post_ok('/notifications',
+              json => { event   => "full moon",
+                        via     => "sms",
+                        channel => '+15555555555' })
+    ->status_is(201)
+    ->json_is('/message' => "notification created");
+
+  # Perform GET requests and look at the responses
+  $t->get_ok('/sunrise')
+    ->status_is(200)
+    ->content_like(qr/ am$/);
+
+  $t->get_ok('/sunset')
+    ->status_is(200)
+    ->content_like(qr/ pm$/);
+
+  # Post a URL-encoded form
+  $t->post_ok('/insurance',
+              form => { name   => 'Jimmy',
+                        amount => '3.000.000',
+                        currency => '€' })
+    ->status_is(200);
+
+  # Use Test::More's like() to check the response
+  like $t->tx->res->dom->at('div#thanks')->text, qr(thank you), "thanks";
+
+  done_testing();
+
+In the rest of this document we'll explore these concepts and others related
+to L<Test::Mojo>.
+
+=head1 CONCEPTS
+
+Essentials every L<Mojolicious> developer should know.
+
+=head2 L<Test::Mojo> At a Glance
+
+The L<Test::More> module bundled with Perl includes several primitive test
+assertions, such as C<ok>, C<is>, C<isnt>, C<like>, C<unlike>, C<cmp_ok>,
+etc. An assertion "passes" if its expression returns a true value. The
+assertion method prints "ok" or "not ok" if an assertion passes or fails
+(respectively).
+
+L<Test::Mojo> supplies additional test assertions organized around the web
+application request/response transaction (transport, response headers,
+response bodies, etc.), and WebSocket communications.
+
+One interesting thing of note: the return value of L<Test::Mojo> object
+assertions is always the test object itself, allowing us to "chain" test
+assertion methods. So rather than grouping related test statements like this:
+
+  $t->get_ok('/frogs');
+  $t->status_is(200);
+  $t->content_like(qr/bullfrog/);
+  $t->content_like(qr/hypnotoad/);
+
+method chaining allows us to connect test assertions that belong together:
+
+  $t->get_ok('/frogs')
+    ->status_is(200)
+    ->content_like(qr/bullfrog/)
+    ->content_like(qr/hypnotoad/);
+
+This makes for a much more I<concise> and I<coherent> testing experience:
+concise because we are not repeating the invocant for each test, and coherent
+because assertions that belong to the same request are syntactically bound in
+the same method chain.
+
+Occsionally it makes sense to break up a test to perform more complex
+assertions on a response. L<Test::Mojo> exposes the entire transaction object
+so you can get all the data you need from a response:
+
+  $t->put_ok('/bees', json => { type => 'worker', name => 'Karl' })
+    ->status_is(202)
+    ->json_has('/id');
+
+  # Pull out the id from the response
+  my $newbee = $t->tx->res->json('/id');
+
+  # Make a new request with data from the previous response
```

Comment:
My point was, as I understand it, that one *shouldn't* pick out data from a response and use it in a subsequent request.  I don't really understand why.  Perhaps it was *why* I was picking data from a response and using it; surely there *is* a good reason to do so.  Perhaps @jberger can help clear this up.  Either way, I think it would be great to explain a recommendation in this regard.  My reason for saying I think it would be good to explain this is that as you've done you explain that you *can* do this, and this leads me to go on to create tests that pick data from a response and use it in a subsequent request, but then I'm advised to not do that, I'm confused because the documentation says that I can...  Hope that makes sense.